src/controller/python/chip/bdx/bdx-transfer.cpp

```diff
@@ -0,0 +1,220 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <controller/python/chip/bdx/bdx-transfer.h>
+
+#include <protocols/bdx/BdxTransferSession.h>
+
+namespace chip {
+namespace bdx {
+namespace {
+
+constexpr uint32_t kMaxBdxBlockSize               = 1024;
+constexpr System::Clock::Timeout kBdxPollInterval = System::Clock::Milliseconds32(50);
+constexpr System::Clock::Timeout kBdxTimeout      = System::Clock::Seconds16(5 * 60);
+
+} // namespace
+
+void BdxTransfer::SetDelegate(BdxTransfer::Delegate * delegate)
+{
+    mDelegate = delegate;
+}
+
+BdxTransfer::~BdxTransfer()
+{
+    if (mData != nullptr)
+    {
+        delete[] mData;
+        mData = nullptr;
+    }
+}
+
+CHIP_ERROR BdxTransfer::AcceptSend()
+{
+    VerifyOrReturnError(mAwaitingAccept, CHIP_ERROR_INCORRECT_STATE);
+    mAwaitingAccept = false;
+
+    TransferSession::TransferAcceptData acceptData;
+    acceptData.ControlMode  = TransferControlFlags::kSenderDrive;
+    acceptData.MaxBlockSize = mTransfer.GetTransferBlockSize();
+    acceptData.StartOffset  = mTransfer.GetStartOffset();
+    acceptData.Length       = mTransfer.GetTransferLength();
+    return mTransfer.AcceptTransfer(acceptData);
+}
+
+CHIP_ERROR BdxTransfer::AcceptReceive(const ByteSpan & data_to_send)
+{
+    VerifyOrReturnError(mAwaitingAccept, CHIP_ERROR_INCORRECT_STATE);
+    mAwaitingAccept = false;
+    mData           = new uint8_t[data_to_send.size()];
+    if (mData == nullptr)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+    memcpy(mData, data_to_send.data(), data_to_send.size());
+    mDataCount = data_to_send.size();
+
+    TransferSession::TransferAcceptData acceptData;
+    acceptData.ControlMode  = TransferControlFlags::kReceiverDrive;
+    acceptData.MaxBlockSize = mTransfer.GetTransferBlockSize();
+    acceptData.StartOffset  = mTransfer.GetStartOffset();
+    acceptData.Length       = mTransfer.GetTransferLength();
+    return mTransfer.AcceptTransfer(acceptData);
+}
+
+CHIP_ERROR BdxTransfer::Reject()
+{
+    VerifyOrReturnError(mAwaitingAccept, CHIP_ERROR_INCORRECT_STATE);
+    mAwaitingAccept = false;
+    return mTransfer.RejectTransfer(StatusCode::kTransferFailedUnknownError);
+}
+
+void BdxTransfer::HandleTransferSessionOutput(TransferSession::OutputEvent & event)
+{
+    assertChipStackLockedByCurrentThread();
+
+    ChipLogDetail(BDX, "Received event %s", event.ToString(event.EventType));
+
+    switch (event.EventType)
+    {
+    case TransferSession::OutputEventType::kInitReceived:
+        mAwaitingAccept = true;
+        mDelegate->InitMessageReceived(this, event.transferInitData);
+        break;
+    case TransferSession::OutputEventType::kStatusReceived:
+        ChipLogError(BDX, "Received StatusReport %x", static_cast<uint16_t>(event.statusData.statusCode));
```

Comment:
Does this error code need to be propagated out somehow?

Reply:
I'm not sure. I don't know if it's useful for the tests to know exactly what status report was received versus knowing that some error occurred. I can plumb it through if you think it would be useful.

Reply:
I think it probably will become useful at some point. Is the status code standard? Can the delegate accept the status code as well, or is the delegate a public API already?